agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp

```diff
@@ -602,7 +602,7 @@ void backhaul_manager::after_select(bool timeout)
 }
 
 bool backhaul_manager::finalize_slaves_connect_state(bool fConnected,
-                                                     std::shared_ptr<SSlaveSockets> pSocket)
+                                                     std::shared_ptr<sRadioInfo> pSocket)
```

Comment:
From a quick look it seems that pSocket is not stored or anything, in which case it's weird to take a shared_ptr instead of a reference as argument (but it was already the case before, so not specific to this PR).

Reply:
You're saying that is should be passed as reference rather than pointer, right? Indeed, that is a fairly general shortcoming in the code currently: in many places, things are passed by shared_ptr just because they happen to be stored as shared_ptr, while really they should be passed as reference (sometimes const reference). But as you say: not for this PR.

Reply:
Yes, that's exactly what I was trying to say :- )

Reply:
I hope that slowly we'll start giving _meaning_ to the types we hold, pass or return.
In particular:
A pointer (shared, unique, raw) - the underlying object may exists or my *not*, and it is OK. In other word we have 0..1 objects behind this pointer.
A reference - we promise that we have _exactly_ one object behind it.
(therefore `A& foo(){ return *a; }` is extremely dangerous and has only rare valid cases)
There are of course many more meanings, like I own it, I transfer ownership, we both own it, you may change it, you may not change and so on.

Reply:
@RanRegev you got this backwards - we want to **pass** by reference to functions, not return reference from functions:
```
bool func()
{
  std::shared_ptr<sVar> var = std::make_shared...
  if (!var) {
    LOG(ERROR) << "Failed to allocate memory";
    return false; 
  }
  ...
  do_something_with_var(*var)
  ...
}

void do_something_with_var(sVar &var);

```

Reply:
No, I didn't get it backward, I was talking generically.
What is "backward" here is 

> we want to *pass* by reference

You never "want to pass", you always pass "what was requested".
The caller does not set the requirements from the parameters, return value or constness, ever. It is the declaration of the function that does this.
And in this manner, the function declaration _states_ that it must have an object under the passed value by setting the parameter with a reference. It can also state "it is fine to have a non-object behind this, I know how to handle it" then it declares a pointer.
functions like:
```
void foo( A* a ){
if ( !a ) return;
/* code with a */
}
```
Are a mistake in the _declaration_.
If you know that your caller has a pointer to A (A*) and therefore you declare your function to request a pointer - this is a mistake. You are not obligated to the way your clients will use you. You are obligated to your needs.
In the example I gave it is better to declare:
```
void foo( A& a ) {
/* code with a */
}
```
and let the caller decide how to call:
```
if ( a ) {
foo( *a ) 
} else {
/* error flow */
}

C++ is a very declarative language, that let you control and emphasis many aspects of your intents and assumptions. I think we should use these abilities more.

Reply:
@tomereli @RanRegev You're both saying the same thing: pass objects by reference, unless it is "optional" (i.e. may be nullptr).